server/controllers/stock/reports/stock/exit_report.js

```diff
@@ -0,0 +1,166 @@
+const {
+  _, db, ReportManager, pdfOptions, STOCK_EXIT_REPORT_TEMPLATE,
+} = require('../common');
+
+const StockExitToPatient = require('./exit/exitToPatient');
+const StockExitToService = require('./exit/exitToService');
+const StockExitToDepot = require('./exit/exitToDepot');
+const StockExitToLoss = require('./exit/exitToLoss');
+
+/**
+ * @method stockExitReport
+ *
+ * @description
+ * This method builds the stock exit report as either a JSON, PDF, or HTML
+ * file to be sent to the client.
+ *
+ * GET /reports/stock/exit
+ */
+function stockExitReport(req, res, next) {
+  let report;
+
+  let data = {};
+  let params = {};
+  const options = req.query;
+  const optionReport = _.extend(options, pdfOptions, {
+    filename : 'REPORT.STOCK.EXIT_REPORT',
+  });
+
+  // set up the report with report manager
+  try {
+    report = new ReportManager(STOCK_EXIT_REPORT_TEMPLATE, req.session, optionReport);
+    params = JSON.parse(req.query.params);
+  } catch (e) {
+    return next(e);
+  }
+
+  /**
+   * get into the data object :
+   * - exitToPatient : stock exit to patient data
+   * - exitToService : stock exit to service data
+   * - exitToDepot : stock exit to depot data
+   * - exitToLoss : stock exit as loss
+   */
+  try {
+    data = collect(params);
+  } catch (e) {
+    return next(e);
+  }
+
+  return fetchDepotDetails(params.depotUuid)
+    .then(depot => {
+      params.depotName = depot.text;
+      return data;
```

Comment:
This is really interesting!  It works ... but I don't think it works the way you think it does ;)

The `collect()` function returns a promise, since it is an [async function](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/async_function).  This means that a clearer representation of what you are doing is shown by renaming `data` to `promise`.
```js
try {
  promise = collect(params);
} catch (e) {
   // This catch() will  never get called since the thing in the try block is a promise
  return next(e);
}

return  fetchDepotDetails(params.depotUuid)
  .then(depot => {
    params.depotName = depot.text;
    return promise;
   })
   .then(/* etc... */);
```
Since the `collect()` returns a promise, the errors will be thrown in the promise chain and cannot be caught by `try/catch`.  This makes your `try/catch` unnecessary.

You could rewrite it like this:
```js
data = collect(params);

return  fetchDepotDetails(params.depotUuid)
  .then(depot => {
    params.depotName = depot.text;
    return data;
   })
   .then(/* etc... */);
```
without changing the logic of the function.  And this is almost equivalent to the following:
```js
return  fetchDepotDetails(params.depotUuid)
  .then(depot => {
    params.depotName = depot.text;
    return collect(params);
   })
   .then(/* etc... */);
```
It's just the `collect()` function fires a bit earlier.

So!  It works, but the `try/catch` is probably unnecessary. :)